osu.Game/Rulesets/Difficulty/DifficultyCalculator.cs

```diff
@@ -107,18 +107,28 @@ public List<TimedDifficultyAttributes> CalculateTimed([NotNull] IEnumerable<Mod>
 
             var skills = CreateSkills(Beatmap, playableMods, clockRate);
             var progressiveBeatmap = new ProgressiveCalculationBeatmap(Beatmap);
+            var difficultyObjects = getDifficultyHitObjects().ToArray();
 
-            foreach (var hitObject in getDifficultyHitObjects())
+            foreach (var obj in difficultyObjects)
             {
-                progressiveBeatmap.HitObjects.Add(hitObject.BaseObject);
+                // Implementations expect the progressive beatmap to only contain top-level objects from the original beatmap.
+                // At the same time, we also need to consider the possibility DHOs may not be generated for any given object,
+                // so we'll add all remaining objects up to the current point in time to the progressive beatmap.
+                for (int i = progressiveBeatmap.HitObjects.Count; i < Beatmap.HitObjects.Count; i++)
```

Comment:
This is a little better but there is still a bit of a problem with this implementation (related to https://github.com/ppy/osu/issues/17964 moreso than to the issue this was originally trying to resolve), although it's only obvious at the point of usage of the method.

Most difficulty calculators skip the first several objects (because they need multiple to work with). This means that for the first object (possibly more, depending on ruleset), the pp calculator will use the `TimedDifficultyCalculator` for the first difficulty object, which may actually be the second or third or fourth _actual_ object.

I'm not sure how much this matters, I guess I'd be willing to let it slide for another day, but given you've already spent time to attempt to fix https://github.com/ppy/osu/issues/17964 with the recent changes too I figure I should at least bring this up before proceeding further...

Reply:
Is there actually an issue with that? Logically I wouldn't think so because it would round to 0pp no matter what.

I'm not sure how to handle that if there is an issue. Perhaps it would be handled in the counter directly rather than clamping to 0.

Reply:
It's probably fine, but I'll give it another check.